### PR TITLE
V1.1.9

### PR DESCRIPTION
**Pull request by devhawala**
This pull-request contains several improvements and changes to Darkstar 1.1.0.0 collected over time.
- The commit message to the first commit in branch v1.1.9 contains a list of the code changes, most important are improvements to the network interface and the ability to boot the guest system from the command line through the configuration file, both on Windows and Unixoids.
- The readme files were updated accordingly.
- The version number 1.1.9.0 was chosen only to keep the version in the forked repository out of the way from possibly new versions in the original repo. So feel free to use a differing version matching the main versioning sequence for Darkstar.
- Finally a disk image for StarOS 5.0 was added, with a matching configuration file; this is also an example for a Darkstar configuration file having all parameters with some comments.

### CODE REVIEW

File: D/Display/DisplayController.cs [hunk 1]
```diff
@@ -171,7 +171,7 @@ private void HorizontalRetraceCallback(ulong skewNsec, object context)
                 {
                     // Normal line : 32 bits of border pattern, 1024 bits of display, 32 bits of border pattern
                     // Border pattern: low byte on lines 4n, 4n+1; high byte on 4n+2, 4n+3.
-                    int patternByte = (effectiveScanline & 0x2) == 0 ? _displayBorder & 0xff : _displayBorder >> 8;
+                    int patternByte = (effectiveScanline & 0x2) == 2 ? _displayBorder & 0xff : _displayBorder >> 8;
```
Comment (jdersch) [inline D/Display/DisplayController.cs:174]: The bug here is more subtle -- changing this here provides the correct pattern alignment for Viewpoint, but breaks the pattern on Interlisp-D.  (Currently Interlisp-D looks fine, VP is broken).  Prefer not changing this now until the root cause can be determined.

File: D/Darkstar.csproj [hunk 2]
```diff
@@ -37,26 +45,40 @@
...
+    <GenerateSerializationAssemblies>Auto</GenerateSerializationAssemblies>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Platform)' == 'x64' ">
+    <Prefer32Bit>true</Prefer32Bit>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
+    <StartAction>Project</StartAction>
+    <StartArguments>-config xde_5.0.cfg</StartArguments>
+    <StartWorkingDirectory>C:\_Hans\Darkstar_NewFeatures\D</StartWorkingDirectory>
+  </PropertyGroup>
```
Comment (livingcomputermuseum) [inline D/Darkstar.csproj:81]: Please remove these extraneous changes to the project

File: D/Display/DisplayController.cs (hunk 1 above)
Comment (livingcomputermuseum) [inline D/Display/DisplayController.cs:174]: The root cause of this issue is more subtle.  Without this change, Interlisp-D's border pattern aligns properly and Viewpoint is misaligned; with this change the opposite will be true.  I suspect the root cause is a timing issue during the start of the frame that I have to dig into.  I'd prefer not to change this right now.  (See also line 216.)

File: D/Ethernet/EthernetController.cs [hunk 3]
```diff
@@ -624,12 +624,14 @@ private void OnHostPacketReceived(MemoryStream data)
...
 
                 if (Log.Enabled) Log.Write(LogComponent.EthernetReceive, "Packet (length {0}) added to pending buffer.", data.Length);
+                //(debug) Console.Write("  -- Packet (length {0}) added to pending buffer -- currently enqueued: {1}\n", data.Length, _pendingPackets.Count);
             }
             else
             {
                 //
                 // Too many queued-up packets, drop this one.
                 //
+                //(debug) Console.WriteLine("Pending buffer full; dropping this packet.");
```
Comment (livingcomputermuseum) [inline D/Ethernet/EthernetController.cs:634]: Please remove commented out debug lines.

File: D/Ethernet/EthernetController.cs [hunk 4]
```diff
@@ -720,6 +722,12 @@ private void ReceiverPollCallback(ulong skewNsec, object context)
                         _inputPacket.Enqueue((ushort)(packetStream.ReadByte() << 8));
                     }
 
+                    // add 2 words to ensure that the ethernet task microcode does not ignore the packet
+                    // (the microcode seems to expect a 4 byte "frame check sequence" at the packet end)
+                    _inputPacket.Enqueue((ushort)0);
+                    _inputPacket.Enqueue((ushort)0);
```
Comment (livingcomputermuseum) [inline D/Ethernet/EthernetController.cs:728]: Are you certain the microcode expects this and that there isn't a different root cause of this issue?  The Ethernet code as it stands works fine under Interlisp-D (tested with the TCP/IP stack) and as far as I know the Ethernet microcode between Interlisp and VP is basically identical.

File: D/Ethernet/NethubInterface.cs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+		private Thread receiverThread = null;
+		
+		private Byte[] sendBuffer = new byte[1026]; // this assumes that there is only ONE processing thread (which is the case for DarkStar!) 
+		
+		public NethubInterface()
+		{
+            client = new TcpClient(Configuration.NetHubHost, Configuration.NetHubPort);
+			stream = client.GetStream();
+			
+			this.localAddress = Configuration.HostID;
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:60]: Un-tabify this file, please (4 spaces per tab).

File: D/Ethernet/NethubInterface.cs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+			this.broadcastAddress = 0x0000FFFFFFFFFFFF;
+		}
+		
+		/// <summary>
+        /// Registers a callback delegate to handle packets that are received.
+        /// </summary>
+        /// <param name="callback"></param>
+        void IPacketInterface.RegisterReceiveCallback(ReceivePacketDelegate callback) {
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:68]: Fix brace style (here and elsewhere in this file.)

File: D/Ethernet/NethubInterface.cs [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        /// </summary>
+        /// <param name="packet"></param>
+        void IPacketInterface.Send(ushort[] packet) {
+        	NetworkStream str = this.stream;
+        	if (str == null) {
+        		return;
+        	}
+        	
+        	int byteLen = packet.Length * 2;
+            //(debug) Console.Write("** sending packed, len = {0} bytes\n", byteLen);
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:86]: Please remove this line (or make it a Log.Write call.)

File: D/Ethernet/NethubInterface.cs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        	} catch (Exception) {
+    			// ignored
+    		}
+        }
+
+        /// <summary>
+        /// Shuts down the encapsulation provider.
+        /// </summary>
+        void IPacketInterface.Shutdown() {
+        	//(debug) Console.WriteLine("** Stopping nethub interface...");
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:118]: Same as above.

File: D/Ethernet/NethubInterface.cs [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+        	}
+	        catch(ThreadInterruptedException)
+	        {
+	            // ignored
+	        }
+        }
+        
+        private int getLenPrefix(NetworkStream str) {
+        	Int32 b1 = this.getByte(str) & 0xFF;
+        	Int32 b2 = this.getByte(str) & 0xFF;
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:181]: Prefer "int" rather than "Int32"

File: D/Program.cs [hunk 10]
```diff
@@ -77,6 +78,10 @@ static void Main(string[] args)
                             }
                             break;
 
+                        case "-start":
+                            doStart = true;
```
Comment (livingcomputermuseum) [inline D/Program.cs:82]: Can you just set Configuration.Start directly here and eliminate doStart?

File: D/UI/ConfigurationDialog.cs [hunk 11]
```diff
@@ -123,13 +125,19 @@ private void PopulateNetworkAdapterList()
             //
             // Populate the list with the interfaces available on the machine, if any.
             //
-            EthernetInterfaceListBox.Enabled = Configuration.HostRawEthernetInterfacesAvailable;
+            EthernetInterfaceListBox.Enabled = true; // Configuration.HostRawEthernetInterfacesAvailable;
```
Comment (livingcomputermuseum) [inline D/UI/ConfigurationDialog.cs:128]: Just remove this line entirely.

File: D/UI/ConfigurationDialog.cs [hunk 12]
```diff
@@ -252,5 +262,22 @@ private void OnClosed(object sender, FormClosedEventArgs e)
...
+            bool isNetHub = (iface.Name == NethubInterface.NETHUB_NAME);
+            txtNetHubHost.Enabled = isNetHub;
+            numNetHubPort.Enabled = isNetHub;
+        }
+
+        private void OnNetHubHostValidating(object sender, System.ComponentModel.CancelEventArgs e)
+        {
+            if (String.IsNullOrEmpty(txtNetHubHost.Text))
+            {
+                MessageBox.Show("The NetHub host can not be empty");
```
Comment (livingcomputermuseum) [inline D/UI/ConfigurationDialog.cs:278]: "cannot"

File: D/UI/DWindow.cs [hunk 13]
```diff
@@ -101,6 +101,21 @@ public DWindow(DSystem system)
...
+			base.OnShown(e);
+			if (firstShown) {
+				firstShown = false;
+				if (Configuration.Start && !_system.IsExecuting)
+	            {
+	                SystemExecutionContext context = new SystemExecutionContext(null, null, null, OnExecutionError);
+	                _system.StartExecution(context);
+	            }
+			}
+		}
```
Comment (livingcomputermuseum) [inline D/UI/DWindow.cs:118]: Fix tabs.

File: D/UI/DWindow.cs [hunk 14]
```diff
@@ -529,6 +550,7 @@ private void UpdateFloppyDriveLabel()
 
         private void PopulateAltBoot()
         {
+        	_system.IOP.MiscIO.AltBoot = Configuration.AltBootMode;
```
Comment (livingcomputermuseum) [inline D/UI/DWindow.cs:553]: Fix tabs.

File: D/readme.txt [hunk 15]
```diff
@@ -57,7 +57,9 @@ use.
 
 If you wish to make use of the emulated Star's Ethernet interface, you
 will need to have WinPCAP installed (on Windows) or libpcap (on the Unix of 
-your choice).  WinPCAP can be downloaded from http://www.winpcap.org/.
+your choice).  WinPCAP can be downloaded from http://www.winpcap.org/.    
+If only Dodo emulated XNS services are to be accessed from Darkstar, a direct
+connection to the Dodo NetHub can be configured instead of using a real network.
```
Comment (livingcomputermuseum) [inline D/readme.txt:62]: Suggest adding a section somewhere covering Dodo (what it is, where to get it, etc...)

File: D/Darkstar.csproj (hunk 2 above)
Comment (livingcomputermuseum) [reply D/Darkstar.csproj:81]: Actually, are any of these changes necessary?

File: D/Ethernet/EthernetController.cs [hunk 16]
```diff
@@ -848,9 +856,12 @@ private void AttachHostEthernet()
             try
             {
 
-                if (Configuration.HostRawEthernetInterfacesAvailable &&
-                    !string.IsNullOrWhiteSpace(Configuration.HostPacketInterfaceName))
-                {
+            	if (Configuration.HostPacketInterfaceName == NethubInterface.NETHUB_NAME) {
+            		_hostInterface = new NethubInterface();
+            		_hostInterface.RegisterReceiveCallback(OnHostPacketReceived);
+            		Console.WriteLine("** connected to nethub");
```
Comment (livingcomputermuseum) [inline D/Ethernet/EthernetController.cs:862]: Would prefer not to have WriteLines in the code (use Log.Write where applicable.)  See other instances in this and other files as well.

File: D/Ethernet/NethubInterface.cs [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+		private TcpClient client = null;
+		private NetworkStream stream = null;
+		
+		private ulong localAddress;
+		private ulong broadcastAddress;
+		
+		private ReceivePacketDelegate receiver = null;
+		private Thread receiverThread = null;
+		
+		private Byte[] sendBuffer = new byte[1026]; // this assumes that there is only ONE processing thread (which is the case for DarkStar!)
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:53]: Member variables should be prefixed with an underscore.

File: D/Ethernet/NethubInterface.cs [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+		
+		/// <summary>
+        /// Registers a callback delegate to handle packets that are received.
+        /// </summary>
+        /// <param name="callback"></param>
+        void IPacketInterface.RegisterReceiveCallback(ReceivePacketDelegate callback) {
+        	this.stopReceiverThread();
+        	this.receiver = callback;
+        	this.receiverThread = new Thread(this.PacketReceiver);
+        	this.receiverThread.Start();
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:72]: No need for "this"

File: D/Ethernet/NethubInterface.cs [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+        void IPacketInterface.Shutdown() {
+        	//(debug) Console.WriteLine("** Stopping nethub interface...");
+        	if (this.stream != null) { this.stream.Close(); }
+        	if (this.client != null) { this.client.Close(); }
+        	this.stopReceiverThread();
+        	this.receiver = null;
+        	Console.WriteLine("** Nethub interface shut down");
+        }
+        
+        private void stopReceiverThread() {
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:126]: PascalCase please

File: D/Ethernet/NethubInterface.cs [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+        	} catch(Exception) {
+        		// ignored
+        	}
+        	if (value < 0) {
+        		throw new ThreadInterruptedException();
+        	}
+        	return (byte)value;
+        }
+        
+        private ulong readAddress(byte[] data, int at) {
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:199]: PascalCase

File: D/Ethernet/NethubInterface.cs [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+        }
+        
+        private int getLenPrefix(NetworkStream str) {
+        	Int32 b1 = this.getByte(str) & 0xFF;
+        	Int32 b2 = this.getByte(str) & 0xFF;
+        	return (b1 << 8) | b2;
+        	
+        }
+        
+        private byte getByte(NetworkStream str) {
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:186]: PascalCase

File: D/Ethernet/NethubInterface.cs [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+                    }
+        		}
+        	}
+	        catch(ThreadInterruptedException)
+	        {
+	            // ignored
+	        }
+        }
+        
+        private int getLenPrefix(NetworkStream str) {
```
Comment (livingcomputermuseum) [inline D/Ethernet/NethubInterface.cs:179]: PascalCase

**Review by livingcomputermuseum — CHANGES_REQUESTED** [review]
See the comments in the code (mostly nitpicky, sorry).  Thanks for submitting this!

**Comment by devhawala** [comment]
Hello,

no problem with pedantic code-reviews :-)

The changes in the new commit should incorporate your comments, hoping not to have overseen something.

For the 2 additional words for incoming packets, i removed the change in Darkstar, relying on the configuration option `ether.useDarkstarWorkaround` in Dodo until the real reason for the lost packet problem is found and fixed. Having thought about it, Dodo may have the issue.

Greetings, Hans

**Review by livingcomputermuseum — APPROVED** [review]
Looks good, thanks!

**Pull request merged**